### PR TITLE
RELATED: TNT-1506 Pivot Table column totals definition

### PR DESCRIPTION
**Pull request by NestorEncinas**
---

Supported PR commands:

| Command                                                 | Description                                                |
| ------------------------------------------------------- | ---------------------------------------------------------- |
| `ok to test`                                            | Re-run standard checks                                     |
| `extended check sonar`                                  | SonarQube tests                                            |
| `extended test - backstop`                              | BackstopJS tests                                           |
| **E2E Cypress tests commands - TIGER**                  |                                                            |
| `extended test - tiger-cypress - isolated <testName>`   | Run isolated tests running against recorded Tiger backend. |
| `extended test - tiger-cypress - record <testName>`     | Create a new recording for isolated Tiger tests.           |
| `extended test - tiger-cypress - integrated <testName>` | Run integrated tests against live backend                  |
| **E2E Cypress tests commands - BEAR**                   |                                                            |
| `extended test - cypress - isolated <testName>`         | Run isolated tests running against recorded Bear backend.  |
| `extended test - cypress - record <testName>`           | Create a new recording for isolated Bear tests.            |
| `extended test - cypress - integrated <testName>`       | Run integrated tests against live backend                  |
| **Compatibility matrix test commands - TIGER Backend**  |                                                            |
| `extended test - matrix-test <AIO_version>`             | Run integrated tests against AIO versions.                 |

`<testName>` in cypress commands is used to filter specfiles. Example, to run record with BEAR backend

-   Against `dashboard.spec.ts` and `drilling.spec.ts`, execute command `extended test - cypress - record dashboard,drilling`
-   Against all specfiles, execute command `extended test - cypress - record` or `extended test - cypress - record *`

`<AIO_version>` in commands is used to start test with multiple AIO instances - each instance in triggered by one jenkins build

-   To run with `master` and `stable`, execute command `extended test - matrix-test master,stable` or `extended test - matrix-test latest`
-   To run with specific version,ex: `2.3.0` and `2.3.1`, execute command `extended test - matrix-test 2.3.0,2.3.1`
-   In case `<AIO_version>` is empty, read versions from file `compTigerVersions.txt` of this repo

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `extended test - backstop` passes
-   [ ] `extended test - tiger-cypress - record` to record new mapping files (Tiger BE)
-   [ ] `extended test - cypress - record` to record new mapping files (Bear BE)
-   [ ] `extended test - tiger-cypress - isolated` passes
-   [ ] `extended test - cypress - isolated` passes
-   [ ] `extended test - tiger-cypress - integrated` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13594/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15244/ : FAILURE in 3m 07s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13603/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15252/ : FAILURE in 8m 43s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13606/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15254/ : FAILURE in 12m 40s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13634/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15283/ : FAILURE in 14m 41s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13639/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15288/ : SUCCESS in 14m 53s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13642/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15291/ : SUCCESS in 16m 18s

File: libs/sdk-ui-pivot/src/impl/cell/cellClass.ts [hunk 1]
```diff
@@ -61,16 +64,29 @@ export function cellClassFactory(
...
             `gd-column-index-${colIndex}`,
             isRowTotal ? "gd-row-total" : null,
+            isColumnTotal ? "gd-column-total" : null,
+            isColumnSubtotal ? "gd-column-subtotal" : null,
             subtotalStyle ? `gd-table-row-subtotal gd-table-row-subtotal-${subtotalStyle}` : null,
             hiddenCell ? "gd-cell-hide s-gd-cell-hide" : null,
             rowSeparator ? "gd-table-row-separator s-gd-table-row-separator" : null,
             isTopPinned && isEmptyCell ? "gd-hidden-sticky-column" : null,
+            // All new totals combinations
+            colAndRowSubtotal ? "gd-table-row-column-subtotal" : null,
```
Comment (pbenes) [inline libs/sdk-ui-pivot/src/impl/cell/cellClass.ts:86]: I see it's not consistent in other fields as well, but what about having at least these all starting with `is`, like `isColAndRowSubtotal` to be aligned with for example the above `isColumnTotal`

File: libs/sdk-ui-pivot/src/impl/structure/tableDescriptorFactory.ts [hunk 2]
```diff
@@ -124,6 +125,13 @@ function groupColumns(
             const groupLevel = columnGroupLevels[level];
             let currentGroup: ScopeCol | undefined = groupLevel.pkToGroup[pathToGroup];
 
+            // TODO: is there a better way?
+            const numberOfTotalHeaders = attributeHeaders.filter((header) =>
```
Comment (pbenes) [inline libs/sdk-ui-pivot/src/impl/structure/tableDescriptorFactory.ts:129]: please consider encapsulating at least to something like 
```
function getTotalInfo(attributeHeaders) {
            const numberOfTotalHeaders = attributeHeaders.filter(isResultTotalHeader(header).length;
            return {
               isTotal: numberOfTotalHeaders > 0 && numberOfTotalHeaders === attributeHeaders.length,
               isSubtotal: numberOfTotalHeaders > 0 && numberOfTotalHeaders !== attributeHeaders.length
            };
}
```
and you can spread the result there later.
Also consider reusing (not sure if possible) as it's there twice.

File: libs/sdk-ui-pivot/src/impl/structure/tableDescriptorTypes.ts [hunk 3]
```diff
@@ -168,6 +168,11 @@ export interface ScopeCol extends TableCol {
      * current group.
      */
     readonly headersToHere: IResultAttributeHeader[];
+
+    // TODO: description
+    readonly isTotal?: boolean;
```
Comment (pbenes) [inline libs/sdk-ui-pivot/src/impl/structure/tableDescriptorTypes.ts:173]: Descriptions

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13644/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15297/ : FAILURE in 2m 26s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13645/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15298/ : SUCCESS in 13m 58s

File: libs/sdk-ui-pivot/src/impl/structure/tableDescriptorTypes.ts (hunk 3 above)
Comment (NestorEncinas) [reply libs/sdk-ui-pivot/src/impl/structure/tableDescriptorTypes.ts:175]: Added.

File: libs/sdk-ui-pivot/src/impl/structure/tableDescriptorFactory.ts (hunk 2 above)
Comment (NestorEncinas) [reply libs/sdk-ui-pivot/src/impl/structure/tableDescriptorFactory.ts]: Moved to function.

File: libs/sdk-ui-pivot/src/impl/cell/cellClass.ts (hunk 1 above)
Comment (NestorEncinas) [reply libs/sdk-ui-pivot/src/impl/cell/cellClass.ts]: Renamed them.

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13649/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/761/ : SUCCESS in 8m 21s

**Comment by NestorEncinas** [comment]
extended test - cypress - isolated

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13652/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/763/ : SUCCESS in 11m 50s

**Review by pbenes — APPROVED** [review]

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR3771

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/2739/ : SUCCESS in 8m 57s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15303/ : SUCCESS in 13m 56s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-bear-isolated-zuul-pipeline/274/ : SUCCESS in 10m 55s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/3896/ : SUCCESS in 11s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul/621/ : FAILURE in 1m 29s (non-voting)
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-tiger-isolated-zuul-pipeline/244/ : SUCCESS in 9m 18s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-compatibility-pipeline/127/ : SUCCESS in 5m 07s (non-voting)

**Pull request merged**